nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/test/java/org/apache/nifi/processors/kudu/TestScanKudu.java

```diff
@@ -0,0 +1,520 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.nifi.processors.kudu;
+
+import java.math.BigDecimal;
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Arrays;
+import org.apache.kudu.ColumnSchema;
+import org.apache.kudu.ColumnTypeAttributes;
+import org.apache.kudu.Type;
+import org.apache.kudu.client.Insert;
+import org.apache.kudu.client.KuduException;
+import org.apache.kudu.client.KuduSession;
+import org.apache.kudu.client.KuduTable;
+import org.apache.kudu.client.PartialRow;
+import org.apache.kudu.test.KuduTestHarness;
+import org.apache.kudu.test.cluster.MiniKuduCluster;
+import org.apache.nifi.reporting.InitializationException;
+import org.apache.nifi.util.MockFlowFile;
+import org.apache.nifi.util.TestRunner;
+import org.apache.nifi.util.TestRunners;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+public class TestScanKudu {
```

Comment:
Considering how the predicate handling had bugs we should add some tests that cover them - not just wether their syntax is valid, but also wether they work as intended.

I created 3 but please cover the other types and predicates as well:
```java
    @Test
    public void testPredicateEqualsFalse() throws Exception {
        List<Boolean> values = Arrays.asList(false, true, false);

        String predicate = "value=false";

        // TODO add proper value
        String expectedContent = "";

        testPredicate(predicate, expectedContent, Type.BOOL, (row, value) -> row.addBoolean("value", value), values);
    }

    @Test
    public void testPredicateGreaterThanInteger() throws Exception {
        List<Integer> values = Arrays.asList(10, 15, 10);

        String predicate = "value>10";

        // TODO This works now, but should be changed, see below
        String expectedContent = "[{\"rows\":[{\"key\":\"2\",\"value\":\"15\"}]}]";

        testPredicate(predicate, expectedContent, Type.INT32, (row, value) -> row.addInt("value", value), values);
    }

    @Test
    public void testPredicateLowerThanOrEqualsFloat() throws Exception {
        List<Float> values = Arrays.asList(3.231F, 3.232F, 3.233F);

        // TODO this test fails because the .232 part of the float doesn't fit the PREDICATES_PATTERN.
        String predicate = "value<=3.232";

        // TODO The "rows" shouldn't be there at the start of each row.
        //  In fact probably shouldn't be there at all as it just makes it harder to chain into other processors.
        String expectedContent = "[{\"rows\":[{\"key\":\"1\",\"value\":\"3.231\"}]},\n" +
                "{\"rows\":[{\"key\":\"2\",\"value\":\"3.232\"}]}]";

        testPredicate(predicate, expectedContent, Type.FLOAT, (row, value) -> row.addFloat("value", value), values);
    }

    private <V> void testPredicate(String predicate, String expectedContent, Type valueType, BiConsumer<PartialRow, V> valueSetter, List<V> values) throws Exception {
        List<ColumnSchema> columns = Arrays.asList(
                new ColumnSchema.ColumnSchemaBuilder("key", Type.INT32).key(true).build(),
                new ColumnSchema.ColumnSchemaBuilder("value", valueType).key(false).build()
        );

        KuduTable kuduTable = kuduScan.getKuduTable(DEFAULT_TABLE_NAME, columns);
        KuduSession kuduSession = kuduScan.kuduClient.newSession();

        values.forEach(value -> {
            Insert insert = kuduTable.newInsert();
            PartialRow row = insert.getRow();

            valueSetter.accept(row, value);

            try {
                kuduSession.apply(insert);
            } catch (KuduException e) {
                throw new RuntimeException(e);
            }
        });

        runner.setProperty(ScanKudu.TABLE_NAME, DEFAULT_TABLE_NAME);
        runner.setProperty(ScanKudu.PREDICATES, predicate);

        runner.setIncomingConnection(false);
        runner.enqueue();
        runner.run(1, false);

        runner.assertTransferCount(ScanKudu.REL_FAILURE, 0);
        runner.assertTransferCount(ScanKudu.REL_SUCCESS, 1);
        runner.assertTransferCount(ScanKudu.REL_ORIGINAL, 1);

        final MockFlowFile flowFile = runner.getFlowFilesForRelationship(ScanKudu.REL_SUCCESS).get(0);
        flowFile.assertContentEquals(expectedContent);
    }
```